### PR TITLE
EZP-28836:  Skipping validation of non-modified fields when content is updated

### PR DESCRIPTION
**Pull request by adamwojs**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-28836](https://jira.ez.no/browse/EZP-28836)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `6.7`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

This PR introduces skipping validation of non-modified fields when content is updated. Otherwise, an editor is unable to publish content if two or more translations contain validation error. 

**TODO**:
- [x] Rebase to 6.7 
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/Repository/Helper/RelationProcessor.php [hunk 1]
```diff
@@ -70,12 +71,16 @@ public function appendFieldRelations(
...
+                        try {
+                            if (!isset($locationIdToContentIdMapping[$locationId])) {
+                                $location = $this->persistenceHandler->locationHandler()->load($locationId);
+                                $locationIdToContentIdMapping[$locationId] = $location->contentId;
+                            }
 
-                        $relations[$relationType][$locationIdToContentIdMapping[$locationId]] = true;
+                            $relations[$relationType][$locationIdToContentIdMapping[$locationId]] = true;
+                        } catch (NotFoundException $e) {
+                            // Ignore missing relation
```
Comment (adamwojs) [inline eZ/Publish/Core/Repository/Helper/RelationProcessor.php]: The side effect of this change is that relations with missing target location will be silently removed.

Comment (alongosz) [reply eZ/Publish/Core/Repository/Helper/RelationProcessor.php]: Let's inject logger `@?logger` (use NullLogger by default) and log this case.

**Comment by alongosz** [comment]
Ref.: integration test: https://github.com/alongosz/ezpublish-kernel/commit/5121bf8fedddac1f6198280c9783cf7c08bad3f0
TODO: negative case when field is updated but wrong link is kept.

**Comment by alongosz** [comment]
Added integration tests for two scenarios:
- Some other field of Content with broken link is updated (ValidationError should not occur)
- Field with broken link is updated keeping broken link (occurs ValidationError)

Also added `eZ\Publish\API\Repository\Tests\PHPUnitConstraint\ValidationErrorOccurs` PHPUnit Constraint that could be useful in many other cases. See 9671a91.

File: eZ/Publish/Core/Repository/Helper/RelationProcessor.php [hunk 2]
```diff
@@ -70,12 +76,18 @@ public function appendFieldRelations(
...
+                                $location = $this->persistenceHandler->locationHandler()->load($locationId);
+                                $locationIdToContentIdMapping[$locationId] = $location->contentId;
+                            }
 
-                        $relations[$relationType][$locationIdToContentIdMapping[$locationId]] = true;
+                            $relations[$relationType][$locationIdToContentIdMapping[$locationId]] = true;
+                        } catch (NotFoundException $e) {
+                            $this->logger->error('Invalid relation: destination location not found', [
+                                'locationId' => $locationId,
+                            ]);
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Helper/RelationProcessor.php]: maybe include `$fieldDefinitionId` also, so it's easier to track Field that refers to missing Location.

File: eZ/Publish/Core/Repository/Tests/Service/Mock/ContentTest.php [hunk 3]
```diff
@@ -5138,7 +5127,71 @@ function (ValueStub $value) use ($emptyValue) {
...
+                    'ger-DE' => 'newValue1-eng-GB',
+                ],
+                'fieldDefinitionId2' => [
+                    'eng-US' => 'newValue2',
+                ],
+            ],
+        ];
+
+        $data = $this->providerForTestUpdateContentNonRedundantFieldSetComplex();
+        for ($i = 0; $i < count($data); ++$i) {
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Tests/Service/Mock/ContentTest.php]: Optimization tip: count will be executed every iteration, but its value is deterministic. It can be moved to a variable outside loop. // Yes really nitpick, it probably saves saved 0.000000001s ;)

File: eZ/Publish/Core/settings/repository.yml [hunk 4]
```diff
@@ -163,3 +164,10 @@ services:
 
     ezpublish.search.background_indexer:
         class: eZ\Publish\Core\Search\Common\BackgroundIndexer\NullIndexer
+
+    ezpublish.repository.relation_processor:
+        class: eZ\Publish\Core\Repository\Helper\RelationProcessor
+        arguments:
+            - "@ezpublish.api.persistence_handler"
+        calls:
+            - ["setLogger", ["@?logger"]]
```
Comment (alongosz) [inline eZ/Publish/Core/settings/repository.yml]: Please use single quotes for new service definition (file is inconsistent anyway due to my reviews ;) )

**Comment by adamwojs** [comment]
PR updated according to @alongosz suggestions.

**Comment by gggeek** [comment]
I see some interesting BC scenario here:

- Article content with id X is valid. Length of, eg. "title" field is 100
- Article contentTpe gets updated, new limitation is set to max length for title to be 80
- Article content with id X is edited, and only Image field is changed

Before the change: edit is rejected
After the change: edit is allowed

Is this what will happen, or am I misunderstanding how this PR is changing the biz logic?

If my understanding is correct, we should *not* allow the edits to go through.
The system should at least cgeck if content class was updated after last content publication and, if so, revalidate all fields

**Comment by gggeek** [comment]
@alongosz sorry, lost in translation, not sure I get the scenario you describe nor the solution.

To me the problem stems from a simple cause: in the GUI the editor edits one translation at a time, whereas the system validates all translations every time (since each version contains all translations).

This is unrelated to the fact that content class definitions can change, making an already published content invalid.

A while ago (around the time kernel v5 was developed), there was a discussion about whether keeping translations "inside" versions made sense or not, as it is wasteful in terms of database and it leads to these weird side effects. Changing this core aspect of the content model is a huge undertaking and, for many reasons, it was not implemented.

So for the moment we have to live with what we have:
- a kernel that thinks that a content version contains all languages
- a gui where by default an editor edits *one language at a time*
- a rest-api / php-api which are based on modifying *one version at a time*, with possibly many languages in it

How can we make this work in an intuitive fashion for both editors and developers?

My opinion is:
- for editors: validate all fields of the *current language*, but not fields of other languages (this is unrelated to checking modified vs. unmodified fields)
- for apis: find the list of translations that are sent along with a new/updated version, and validate those languages but not any language that is not updated. This is more flexible than the GUI but still follows the same principle

I personally do not see the need for a system that allows editing parts of content and allow publication while other parts of the same content are in an invalid state. Maybe you can elaborate on this scenario ?

**Comment by emodric** [comment]
> I personally do not see the need for a system that allows editing parts of content and allow publication while other parts of the same content are in an invalid state.

I agree with this. The system should make sure that the content has a valid state, according to rules from content type, at time of publication.

**Comment by gggeek** [comment]
@andrerom agree with the scenario. That's why I asked in original comment if the idea was to not-validate only unmodified fields in other-languages or also to not-validate unmodified fields in current language ;-)

**Comment by andrerom** [comment]
Well "current language" is a bit misconception when api allows several, and this is where we are currently fixing it.

But indeed we could try to validate _all_ fields of _all_ user changed translations, thus avoiding users of API circumventing validation by omitting certain fields.

WDYT @alongosz / @adamwojs ?

**Comment by alongosz** [comment]
> But indeed we could try to validate all fields of all user changed translations, thus avoiding users of API circumventing validation by omitting certain fields.

Isn't that the initial fix @adamwojs suggested?

> I personally do not see the need for a system that allows editing parts of content and allow publication while other parts of the same content are in an invalid state. Maybe you can elaborate on this scenario ?

@gggeek I just wondered if we should validate things that user did not change. I didn't have in mind anything specific, just the idea that user might not use Admin UI for that but his custom solution that calls directly the API. In this case there's a lot of possibilities.

Let's imagine someone uses Content to represent some Document with well defined Fields. One user can edit some Fields, the other user can edit different Fields (I know that's possible only on custom UI, but still - possible). 

In that scenario if validation fails on fields editable by different users there's no real possibility to save that document at all.

But maybe I'm overthinking this :)

Having said that I'm fine with both solutions - the one with validating only updated fields and with the one validating all fields in updated translation(s).

**Comment by gggeek** [comment]
> But maybe I'm overthinking this :)

I would agree ;-)

Before we allow different users to edit only some fields within the document, we would need to have per-field policies implemented (leaving access control to the hypothetical custom-gui is a risky proposition), as well as maybe better tracing of who-did-what-when, and drafts that can be shared between users.

One other way to achieve this could be instead to have a master-document as composition of multiple content objects, each with its own permissions and versions. In that case we would need to have the custom gui taking care of making sense of all the different versions and presenting it as a coherent set of data, which is not an easy task (but I know it has been done by multiple customers already in the past)

Long story short: worthwile goal, but needs much more work than what this ticket would allow by itself...

**Comment by gggeek** [comment]
> Note: It might be a blocker/complication for future feature though: Inline Editing

True.
But that one might warrant as well a separate API, which is not `publishVersion` but `updateField` instead.

Because the side effects of storing one version for each filed update could be significant:
- version number would be spammed. Either we end up with a huge db, or we have to trim automatically saved versions to max. N (but then we need to allow editors to tag important versions, a bit like google allows to do, to avoid them getting eliminated automatically)
- current publishing operation is heavy. Either we need async js, or we can not achieve smooth experience
- probably more that I can't think of right now

**Comment by andrerom** [comment]
> Isn't that the initial fix @adamwojs suggested?

afaik no, I think it was only validating the fields in the initial language of the struct, but the struct might contain fields in several languages, so think the suggestion was to validate ALL fields, in ALL struct present languages.

File: eZ/Publish/Core/Repository/ContentService.php [hunk 5]
```diff
@@ -1187,7 +1187,16 @@ public function updateContent(APIVersionInfo $versionInfo, APIContentUpdateStruc
         }
 
         $mainLanguageCode = $content->contentInfo->mainLanguageCode;
-        $languageCodes = $this->getLanguageCodesForUpdate($contentUpdateStruct, $content);
+        if ($contentUpdateStruct->initialLanguageCode === null) {
+            $contentUpdateStruct->initialLanguageCode = $mainLanguageCode;
```
Comment (andrerom) [inline eZ/Publish/Core/Repository/ContentService.php]: Is this a separate bug fix?

File: eZ/Publish/Core/Repository/ContentService.php [hunk 6]
```diff
@@ -1187,7 +1187,16 @@ public function updateContent(APIVersionInfo $versionInfo, APIContentUpdateStruc
...
+        }
+
+        $allLanguageCodes = $this->getLanguageCodesForUpdate($contentUpdateStruct, $content);
+        foreach ($allLanguageCodes as $languageCode) {
+            $this->persistenceHandler->contentLanguageHandler()->loadByLanguageCode($languageCode);
```
Comment (andrerom) [inline eZ/Publish/Core/Repository/ContentService.php]: Seems to be purely for validation of the language codes, this was not done in the past was it?

**Comment by mnocon** [comment]
@adamwojs it works for the scenario described in the ticket, but if I do the following steps:

1. Setup more than one language (eng-GB, pol-PL)
2. Create new Article - leave the non required field "author" empty
3. Translate it to Polish (again, leave the non required field "author" empty)
4. Change Content Type settings so that author field is required
5. Try to publish another version of Article (in any language)

Whichever translation I choose I get an error when publishing, with message like:
```
{"ErrorMessage":{"_media-type":"application\/vnd.ez.api.ErrorMessage+json","errorCode":400,"errorMessage":"Bad Request","errorDescription":"Content fields did not validate","errorDetails":{"fields":[{"_fieldTypeId":153,"errors":[{"type":"empty","message":"Value for required field definition 'author' with language 'eng-GB' is empty"}]}]}}}
```
about the other translation than the one I'm editing.

I believe it shouldn't be that way after changes or is editing Content Type different?

File: eZ/Publish/Core/Repository/ContentService.php (hunk 6 above)
Comment (alongosz) [inline eZ/Publish/Core/Repository/ContentService.php]: Hmm, AFAICS this existed before, so maybe not an issue, but why is it needed if we don't use result?

Comment (adamwojs) [reply eZ/Publish/Core/Repository/ContentService.php]: IMHO intention of this call in the original code was to check if a language exists. Ref:  https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Repository/ContentService.php#L1314

Comment (alongosz) [reply eZ/Publish/Core/Repository/ContentService.php]: Hmm, yep, that makes sense. Still weird to see it here. But let's keep it.

**Comment by adamwojs** [comment]
@mnocon I confirm possible stalemate in case of changing `required` attribute on Content Type definition.  

https://github.com/adamwojs/ezpublish-kernel/blob/9c970e1d5221d357ae784d4cda1b4221353f3394/eZ/Publish/Core/Repository/ContentService.php#L1246-L1251 should be skipped if the language is not updated.

**Comment by andrerom** [comment]
@alongosz wdyt? We should adapt that part too I guess to apply the same language logic (?)

**Comment by alongosz** [comment]
@andrerom @adamwojs yes, that makes sense. We need to handle it uniformly in all relevant places.

@adamwojs I have only one request - could you cleanup commits? If it's possible to divide changed code into logical separate pieces that would be good for review. If not then just squash relevant fixups.

*(Usually we don't recommend that but in this case solution changed significantly from the initial one).*

File: eZ/Publish/Core/Repository/ContentService.php [hunk 7]
```diff
@@ -1330,6 +1340,34 @@ public function updateContent(APIVersionInfo $versionInfo, APIContentUpdateStruc
...
+        $languageCodes = [
+            $contentUpdateStruct->initialLanguageCode => true,
+        ];
+
+        foreach ($contentUpdateStruct->fields as $field) {
+            if ($field->languageCode === null || isset($languageCodes[$field->languageCode])) {
+                continue;
+            }
+
+//            dump($field->fieldDefIdentifier, $field->languageCode);
```
Comment (andrerom) [inline eZ/Publish/Core/Repository/ContentService.php]: this and the one below

**Comment by adamwojs** [comment]
I've added integration test for the case discovered by @mnocon.

File: eZ/Publish/API/Repository/Tests/ContentServiceTest.php [hunk 8]
```diff
@@ -1476,6 +1477,60 @@ public function testUpdateContentThrowsContentFieldValidationException()
         /* END: Use Case */
     }
 
+    /**
+     * Test for the updateContent() method.
+     *
+     * @see \eZ\Publish\API\Repository\ContentService::updateContent()
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/ContentServiceTest.php]: Should be `@covers`.

File: eZ/Publish/API/Repository/Tests/ContentServiceTest.php [hunk 9]
```diff
@@ -1476,6 +1477,60 @@ public function testUpdateContentThrowsContentFieldValidationException()
...
+     * @depends eZ\Publish\API\Repository\Tests\ContentServiceTest::testUpdateContent
+     */
+    public function testUpdateContentIgnoreIsRequiredOnNonUpdatedLanguages()
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/ContentServiceTest.php]: I got lost when reading that method name :)
Maybe `testUpdateContentValidatorIgnoresRequiredFieldsOfNotUpdatedLanguages`

File: eZ/Publish/Core/Repository/Tests/Service/Mock/RelationProcessorTest.php [hunk 10]
```diff
@@ -259,6 +263,58 @@ public function testAppendFieldRelationsLocationMappingWorks()
         );
     }
 
+    public function testAppendFieldRelationsLogMissingLocations()
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Tests/Service/Mock/RelationProcessorTest.php]: Nitpick: `Logs` because method is in singular form.

**Comment by adamwojs** [comment]
PR updated according to @alongosz suggestions.
